cl/donate/api_views.py

```diff
@@ -0,0 +1,215 @@
+import json
+from http import HTTPStatus
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ObjectDoesNotExist
+from django.http import HttpResponse
+from rest_framework import mixins, serializers, viewsets
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from cl.donate.api_permissions import AllowNeonWebhook
+from cl.donate.models import NeonMembership, NeonWebhookEvents
+from cl.lib.neon_utils import NeonClient
+
+
+class NeonMembershipWebhookSerializer(serializers.Serializer):
+    eventTrigger = serializers.CharField(required=True)
+    eventTimestamp = serializers.DateTimeField(required=True)
+    data = serializers.DictField()
+
+
+class MembershipWebhookViewSet(
+    mixins.CreateModelMixin, viewsets.GenericViewSet
+):
+    permission_classes = [AllowNeonWebhook]
+    serializer_class = NeonMembershipWebhookSerializer
+    queryset = NeonMembership.objects.all()
+
+    def create(self, request: Request, *args, **kwargs):
+        """
+        Processes membership webhooks received from Neon CRM.
+
+        This method handles POST requests containing JSON data from Neon CRM.
+        It extracts relevant information from the data and updates the
+        corresponding membership record in the database.
+
+        Args:
+            request (Request): A request object containing the webhook data.
+
+        Returns:
+            Response: A HttpResponse object with status code 200 if successful,
+            or another appropriate response if an error occurs.
+        """
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        webhook_data = serializer.validated_data
+        self._store_webhook_payload(webhook_data)
+        match webhook_data["eventTrigger"]:
+            case "createMembership" | "editMembership":
+                self._handle_membership_creation_or_update(webhook_data)
+            case "deleteMembership":
+                self._handle_membership_deletion(webhook_data)
+            case "updateMembership":
+                pass
```

Comment:
Why do we skip this? Is it the same as `editMembership`? Maybe a comment here?

Reply:
> Is it the same as editMembership?

No, It's not. I only received the `updateMembership` webhook after upgrading my membership from my account dashboard, which is separate from the admin dashboard.

I noticed that the upgrade process triggers two webhooks almost simultaneously.:

1. One to update the old membership. The `eventTrigger` of this webhooks is  `updateMembership` and the payload contains details of the previous membership record but with an updated 'termEndDate' field. As a result, the status of the previous membership will be changed to expired in Neon.

2. One to create a new membership. The `eventTrigger` of this webhook is `createMembership` and the payload has all the details of the new membership(new `membershipId`, new `level`, and new `termEndDate`).

> Why do we skip this?

I decided to skip the `updateMembership` webhook to prevent a race condition during the upgrade process. We only need the latest information to update the membership record in our database. However, there's a chance that the `createMembership` webhook might be processed before the `updateMembership` webhook, leading to data integrity issues.

Reply:
Got it. Can you make a comment about this in the code, please to explain why we ignore this? I guess the easiest thing is:

```
# See: https://github.com/freelawproject/courtlistener/pull/3468#discussion_r1433398175
```

:)